lib/network/rest/server.js

```diff
@@ -0,0 +1,332 @@
+     *
+     * @param {http.IncomingMessage} req - HTTP request object
+     * @param {http.ServerResponse} res - HTTP response object
+     * @param {Logger} log - logger object
+     * @return {undefined}
+     */
+    _onPut(req, res, log) {
+        const urlObj = url.parse(req.url);
+        const pathInfo = utils.explodePath(urlObj.path);
+        if (!pathInfo) {
+            return sendError(res, 403, 'invalid URL');
+        }
+        if (pathInfo.service !== constants.dataFileURL) {
+            return sendError(res, 403,
+                             `unsupported service '${pathInfo.service}'`);
+        }
+        if (pathInfo.key !== undefined) {
+            // note: we may implement rewrite functionality by
+            // allowing a key in the URL, though we may still provide
+            // the new key in the Location header to keep immutability
+            // property and atomicity of the update (we would just
+            // remove the old object when the new one has been written
+            // entirely in this case, saving a request over an
+            // equivalent PUT + DELETE).
+            return sendError(res, 403, 'PUT url cannot contain a key');
+        }
+        const contentLength = req.headers['content-length'];
+        if (contentLength === undefined) {
+            return sendError(res, 400, 'missing Content-Length');
+        }
+        const size = Number.parseInt(contentLength, 10);
+        if (isNaN(size)) {
+            return sendError(res, 400, 'bad Content-Length');
+        }
+        this.dataStore.put(req, size, log, (err, key) => {
+            if (err) {
+                return sendError(res, err.code, err.message);
+            }
+            log.debug('sending back 201 response');
+            res.setHeader('Location', `${constants.dataFileURL}/${key}`);
+            setContentLength(res, 0);
+            res.writeHead(201);
+            return res.end();
+        });
+        return undefined;
+    }
+
+    /**
+     * Handler for GET requests
+     *
+     * @param {http.IncomingMessage} req - HTTP request object
+     * @param {http.ServerResponse} res - HTTP response object
+     * @param {Logger} log - logger object
+     * @return {undefined}
+     */
+    _onGet(req, res, log) {
+        const urlObj = url.parse(req.url);
+        const pathInfo = utils.explodePath(urlObj.path);
+        if (!pathInfo) {
+            return sendError(res, 403, 'invalid URL');
+        }
+        if (pathInfo.service !== constants.dataFileURL) {
+            return sendError(res, 403,
+                             `unsupported service '${pathInfo.service}'`);
+        }
+        if (pathInfo.key === undefined) {
+            return sendError(res, 403, 'URL is missing key');
+        }
+        const rangeHeader = req.headers.range;
+        let rangeSpec = undefined;
+        if (rangeHeader !== undefined) {
+            try {
+                rangeSpec = parseRangeSpec(rangeHeader);
```

Comment:
Just came back to me after more reading: the numbers parsed are necessarily valid positive numbers because already checked by the regexp. (We don't check that they don't overflow though).